### PR TITLE
fix(training): Explicit Batch invariance

### PR DESCRIPTION
**Pull request by HCookie**
## Description

Add explicit batch normalisation to the loss function

## Type of Change

-   [x] Bug fix (non-breaking change which fixes an issue)
-   [ ] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] Documentation update

## Code Compatibility

-   [x] I have performed a self-review of my code

### Code Performance and Testing

-   [x] I have added tests that prove my fix is effective or that my feature works
-   [x] I ran the [complete Pytest test](https://anemoi.readthedocs.io/projects/training/en/latest/dev/testing.html) suite locally, and they pass
-   [ ] I have tested the changes on a single GPU
-   [ ] I have tested the changes on multiple GPUs / multi-node setups
-   [ ] I have run the Benchmark Profiler against the old version of the code
-   [ ] If the new feature introduces modifications at the config level, I have made sure to update Pydantic Schemas and default configs accordingly

### Dependencies

-   [x] I have ensured that the code is still pip-installable after the changes and runs
-   [x] I have tested that new dependencies themselves are pip-installable.
-   [x] I have not introduced new dependencies in the inference portion of the pipeline

### Documentation

-   [x] My code follows the style guidelines of this project
-   [x] I have updated the documentation and docstrings to reflect the changes
-   [x] I have added comments to my code, particularly in hard-to-understand areas

### CODE REVIEW

File: training/src/anemoi/training/losses/base.py [hunk 1]
```diff
@@ -111,10 +112,18 @@ def scale(
         return x[subset_indices] * scaler[subset_indices]
 
     def reduce(self, out: torch.Tensor, squash: bool = True) -> torch.Tensor:
+        """Reduce the out of the loss.
+
+        If `squash` is True, the last dimension is averaged.
+        Then averaged over the batch size, and finally summed over the batch size, ensemble and grid dimensions.
```
Comment (mchantry) [inline training/src/anemoi/training/losses/base.py]: `averaged over the batch size, and finally summed over the batch size`
should be
`normalised by the batch size, ...`

**Comment by mchantry** [comment]
How does this interact with multi-GPU batch sizes? Is this part of the all-reduce an average?

File: training/src/anemoi/training/losses/base.py (hunk 1 above)
Comment (HCookie) [reply training/src/anemoi/training/losses/base.py]: Fixed in [cb8c461](https://github.com/ecmwf/anemoi-core/pull/318/commits/cb8c461d4638cb6c751d982ef8fd666498eb2447)

**Comment by HCookie** [comment]
@mchantry 
> Is this part of the all-reduce an average?

As far as I know, the DDP handles the averaging across distributed models and nodes. It will expect the loss to be normalised over the 'mini' - batch

**Comment by mc4117** [comment]
Here are some mlflow runs showing same performance pre loss scaling PR and post fix

Batch size 4
[https://mlflow.ecmwf.int/#/metric?runs=[%229d26d9b95dba4555802b2ca5852cf680%22,%22467e94edf0cb40f8a61ec549fb32445e%22]&metric=%22epoch%22&experiments=[%2254%22]&plot_metric_keys=%5B%22train_mse_loss_step%22%2C%22train_wmse_step%22%5D&plot_layout={%22autosize%22:true,%22xaxis%22:{},%22yaxis%22:{}}&x_axis=relative&y_axis_scale=linear&line_smoothness=1&show_point=false&deselected_curves=[]&last_linear_y_axis_range=[]](https://mlflow.ecmwf.int/#/metric?runs=%5B%229d26d9b95dba4555802b2ca5852cf680%22,%22467e94edf0cb40f8a61ec549fb32445e%22%5D&metric=%22epoch%22&experiments=%5B%2254%22%5D&plot_metric_keys=%5B%22train_mse_loss_step%22%2C%22train_wmse_step%22%5D&plot_layout=%7B%22autosize%22:true,%22xaxis%22:%7B%7D,%22yaxis%22:%7B%7D%7D&x_axis=relative&y_axis_scale=linear&line_smoothness=1&show_point=false&deselected_curves=%5B%5D&last_linear_y_axis_range=%5B%5D)

Batch size 2
[https://mlflow.ecmwf.int/#/compare-runs?runs=[%227996170d02294357bcbec9b7f3007237%22,%226589ceede8de407da89250a3125b959a%22]&experiments=[%2254%22]](https://mlflow.ecmwf.int/#/compare-runs?runs=%5B%227996170d02294357bcbec9b7f3007237%22,%226589ceede8de407da89250a3125b959a%22%5D&experiments=%5B%2254%22%5D)

**Comment by mc4117** [comment]
Before this fix, there was no batch size invariance as shown here:

https://mlflow.ecmwf.int/#/metric?runs=[%22467e94edf0cb40f8a61ec549fb32445e%22,%226589ceede8de407da89250a3125b959a%22,%2275d102b37e30428c999beb66ada83008%22,%223a48ef1e6b354cd8ae21f81422cbb91c%22]&metric=%22system%2Fcpu_utilization_percentage%22&experiments=[%2254%22]&plot_metric_keys=%5B%22train_wmse_step%22%2C%22train_mse_loss_step%22%5D&plot_layout={%22autosize%22:true,%22xaxis%22:{%22autorange%22:true,%22type%22:%22linear%22},%22yaxis%22:{}}&x_axis=step&y_axis_scale=linear&line_smoothness=1&show_point=false&deselected_curves=[]&last_linear_y_axis_range=[]

**Comment by ssmmnn11** [comment]
> batch size invariance

Is this summary correct?

https://github.com/ecmwf/anemoi-core/pull/52 removed (broke) batch size invariance

this fix restores it, we get the same loss as we had before https://github.com/ecmwf/anemoi-core/pull/52?

File: training/src/anemoi/training/losses/base.py [hunk 2]
```diff
@@ -110,11 +118,42 @@ def scale(
...
+        -------
+        torch.Tensor
+            Reduced output tensor
+        """
         if squash:
-            out = self.avg_function(out, dim=-1)
+            out = self.avg_function(out, dim=TensorDim.VARIABLE)
 
-        return self.sum_function(out, dim=(0, 1, 2))
+        grid_summed = self.sum_function(out, dim=(TensorDim.GRID))
```
Comment (mchantry) [inline training/src/anemoi/training/losses/base.py:149]: Could you add a comment to explain that this is a sum because the normalisation is handled in the GRID weighting?
